vault/eventbus/bus.go

```diff
@@ -217,9 +226,51 @@ func (bus *EventBus) Subscribe(ctx context.Context, ns *namespace.Namespace, pat
 	// add info needed to cancel the subscription
 	asyncNode.pipelineID = eventlogger.PipelineID(pipelineID)
 	asyncNode.cancelFunc = cancel
+	go bus.replayEvents(asyncNode)
 	return asyncNode.ch, asyncNode.Close, nil
 }
 
+func (bus *EventBus) copyRecentEvents() []*logical.EventReceived {
```

Comment:
WDYT about taking some inspiration from the [consul replay buffer](https://github.com/hashicorp/consul/blob/db69cd6f659726cce0602f58322e35527f0fe985/agent/consul/stream/event_buffer.go#L10-L61), keeping the recent events as an append-only linked list so that we can safely traverse it in multiple threads without taking a lock or copying? I had a bit of a go at what that could look like [here](https://github.com/hashicorp/vault/commit/ee05bfdb57836b378f6884230a61a9d1213ac1bb#diff-349ec5d48b5fc184ac02a84d476cbfc650330f23371d34cbdfd01f1c7a95ba29R67-R104).

Reply:
I thought about this. It simplifies a few things, but means more allocations in the hot path and more GC, so I went with the slice-backed `deque`, which should be a fixed size array.

I think using the slice-backed `deque` is the right approach if we expect sending events to be more common than new subscriptions.